[HACK week] Enables favorite products filter feature

selanthiraiyan opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #12274 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description

Enables the HACK week feature - option to filter favorite products in product lists screen and product selector screen.

⚠️ Don't merge until we get a go-ahead here. pe5sF9-3ag-p2#comment-4157

Changes
- Enables feature flag.
- Adds missing tracking.
- Removed unnecessary MainActor usage to remove warnings. 


## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

#### Prerequisites
- Use a Woo store with more than 100 products. 
- You can create a JN store and generate 100+ products using Woo Smooth generator. 
- Using the smooth generator helps to generate products with different types, status and category. (Helps with testing filters)
- Follow instructions from [this PR](https://github.com/woocommerce/woocommerce-ios/pull/12284) and mark a few products as favorites. 

#### Products list
- Login into the app if needed
- Navigate to Products tab
- Tap "Filter"
- Validate that you see the "Favorite Products" filter.
- Select the "Favorite Products" filter -> Switch to "Favorite Products" from "Any"
- Tap "Show Products"
- Xcode logs should track the following event with the new `favorite_products` value in `filters` propoerty. 
```
🔵 Tracked product_filter_list_show_products_button_tapped, properties: [is_wpcom_store: false, blog_id: 210109692, site_url: https://yoursite..com, store_id: 121212-7fa0cb61a383, was_ecommerce_trial: false, plan: jetpack_security_daily, filters: favorite_products, source: products_tab]
```
- Validate that favorite products are loaded correctly.
- Apply other filters and ensure that they work as before. 

#### Product selector
- Login into the app if needed
- Navigate to Orders tab
- Tap "+" -> "Add products"
- Tap "Filter"
- Validate that you see the "Favorite Products" filter.
- Select the "Favorite Products" filter -> Switch to "Favorite Products" from "Any"
- Tap "Show Products"
- Xcode logs should track the following event with the new `favorite_products` value in `filters` propoerty. 
```
🔵 Tracked product_filter_list_show_products_button_tapped, properties: [source: order_form, plan: jetpack_security_daily, filters: favorite_products, blog_id: 210109692, site_url: https://yoursite.com, is_wpcom_store: false, store_id: 12121212-7fa0cb61a383, was_ecommerce_trial: false]
```
- Validate that favorite products are loaded correctly.
- Apply other filters and ensure that they work as before. 

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
- Tested the product filtering works when applying the favorite product filter.
- Tested that product filtering works correctly when applying multiple filters along with the favorite product filter.
- Tested that product filtering works correctly in both product selector and product list screens.
- Tested that adding filters tracks the correct tracks event with correct properties. 

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

#### Gif

iPhone
| Product list | Product selector |
|--------|--------|
| ![Simulator Screen Recording - iPhone 15 Pro Max - 2024-09-20 at 13 38 31](https://github.com/user-attachments/assets/35aaa076-5e52-4fcd-9a51-29d9c7ac130a) | ![Simulator Screen Recording - iPhone 15 Pro Max - 2024-09-20 at 13 38 47](https://github.com/user-attachments/assets/87ed8d76-605e-4295-98e5-03e585bc74a9) |

iPad
| Product list | Product selector |
|--------|--------|
| ![Simulator Screen Recording - iPad Pro 13-inch (M4) - 2024-09-20 at 13 42 02](https://github.com/user-attachments/assets/87d54880-8640-4455-8746-ac795815db8d) | ![Simulator Screen Recording - iPad Pro 13-inch (M4) - 2024-09-20 at 13 43 33](https://github.com/user-attachments/assets/1816ddab-a79f-4a4d-8eb0-48606145f7b2) |


#### Video recording

| Product list | Product selector |
|--------|--------|
| https://github.com/user-attachments/assets/65e16089-5594-4497-974b-aa52a766ee20 | https://github.com/user-attachments/assets/07fb837d-37a1-41ca-8fc2-c357f0588971 | 


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

selanthiraiyan committed:
Track favorite product filter in analytics event.

selanthiraiyan committed:
Remove unnecessary MainActor usage to remove Xcode warnings.

selanthiraiyan committed:
Remove `favoriteProducts` feature flag.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

selanthiraiyan committed:
Add release notes for favotire products change.

selanthiraiyan committed:
Remove feature flag from tests.

selanthiraiyan committed:
Inject mock stores to fix unit tests failing due to main thread error.

itsmeichigo reviewed (commented):
Hi Sharma, I thought we wanted to keep the feature flags and only remove them after a few releases? That would make it easier to backtrack in case something breaks.

One more thing: I think the filter in the product selector affects the list in Product tab as well:


https://github.com/user-attachments/assets/9537875a-05b9-4a67-99a8-e955a6873d56

This seems to be the issue for all fitlers not just the favorite one, so we should probably log a separate issue to fix.

selanthiraiyan committed:
Revert "Remove feature flag from tests."

This reverts commit 04edf72c5f63ccdb02f37060d872896979d9a8b5.

selanthiraiyan committed:
Revert "Remove `favoriteProducts` feature flag."

This reverts commit 5a3185d64d8937980bf54c939681ffcf289af13f.

selanthiraiyan committed:
Enable favorite products feature flag.

selanthiraiyan commented:
> One more thing: I think the filter in the product selector affects the list in Product tab as well:

> This seems to be the issue for all fitlers not just the favorite one, so we should probably log a separate issue to fix.

Good catch, Huong! I logged an issue https://github.com/woocommerce/woocommerce-ios/issues/14624

selanthiraiyan committed:
Merge branch 'trunk' into feat/12274-release

selanthiraiyan commented:
Thanks for the review, Huong! 🙇 

> I thought we wanted to keep the feature flags and only remove them after a few releases? That would make it easier to backtrack in case something breaks.

I agree, Huong. I thought removing the feature flag would be cleaner due to the minimal usage of the feature flag in code. 
- Reverted the changes and enabled feature flag in https://github.com/woocommerce/woocommerce-ios/pull/14597/commits/b593f4b77eb0af37c8772d775aceabf30f23d38b. 
- I have added a subtask to remove feature flag later. https://github.com/woocommerce/woocommerce-ios/issues/12274

The PR is ready for another round of review.

itsmeichigo reviewed (approved):
Thank you for the updates!

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/238">21.8</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

selanthiraiyan committed:
Merge branch 'trunk' into feat/12274-release

selanthiraiyan committed:
Move release notes to latest version.

selanthiraiyan committed:
Merge branch 'trunk' into feat/12274-release

selanthiraiyan commented:
Merging this PR to release this feature based on this internal discussion - pe5sF9-3ag-p2#comment-4406

After merging recent changes from `trunk`, I have tested the changes to ensure that marking/removing favorites and filtering  favorite products work as expected.
